### ISSUE TITLE
How about a Set equivalent?

### ISSUE THREAD

**Issue by Herteby** [pzp1997/assoc-list#1]
Any plans on adding a corresponding `Set` to this package? :slightly_smiling_face: 
It could essentially just be `type Set a = Set (Dict a ())`

**Comment by erlandsona** [comment]
Actually you can try it out too if you want!
@pzp1997 hope it's alright I've copy/pasted the docs for now.
https://package.elm-lang.org/packages/erlandsona/assoc-set/latest/

I've never published a library before so this was really cool!
Other than updating the docs I obviously I could just comment out the 9 failing tests too...
Curious what tips y'all have for measuring perf and such?

**Comment by pzp1997** [comment]
> Any plans on adding a corresponding `Set` to this package? 🙂

Yeah, I suppose I could do this if people have a good use for it. I did not do it initially because I saw a more immediate need for a `Dict` substitute. I felt that the `Set` API was simple enough that people could make their own `Set`-like data structures on an ad-hoc basis. Additionally, there is already a package named [Chadtech/unique-list](https://package.elm-lang.org/packages/Chadtech/unique-list/latest/) that I thought accomplishes something pretty similar (although it doesn't exactly implement the `Set` API).

> It could essentially just be `type Set a = Set (Dict a ())`

We _could_ do it this way, but I am curious if we can get better perf by keeping the implementations of `Dict` and `Set` separate. I started working on separate implementations of the `Set` API earlier this week but haven't had time to run perf tests yet. (Surprisingly, elm/core actually implements `Set` exactly as you suggested so maybe it would be fine for us too.)

> I'm not sure how to publish and 9 of the tests I copied from elm/core are failing but I'm sure there's some details with utilizing assoc-list that I just need to iron out before publishing...

I haven't had a chance to run the tests for your package yet, but if I had to take a guess the tests are failing because my package and elm/core have different invariants regarding the ordering of entries. You should modify the failing tests to account for this difference.

> I've never published a library before so this was really cool!

Congrats on publishing your first library! For the sake of cohesiveness, though I would prefer to keep the `Dict` and `Set` implementations together in this library. Are you interested in working on this further and making a pull request on my repo?

> Curious what tips y'all have for measuring perf and such?

I really like [elm-explorations/benchmark](https://package.elm-lang.org/packages/elm-explorations/benchmark/latest/) for measuring perf.

**Comment by erlandsona** [comment]
Yeh happy to PR my fork alternatively! I just did it initially for the experience. elm packages really is wonderful!
I'll have to check out elm-explorations/benchmark! Thanks for tips!

**Comment by amitaibu** [comment]
https://github.com/Gizra/elm-all-set is based on this package
